File: thinking-categorically/functor/contravariant-hom-functor/contravariant-hom-functor.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Contravariant Hom functor implementation
public interface IContravariant<in A>
{
    IContravariant<B> Contramap<B>(Func<B, A> f);
}

// Predicate contravariant functor
public class Predicate<A> : IContravariant<A>
{
    private readonly Func<A, bool> predicate;

    public Predicate(Func<A, bool> predicate)
    {
        this.predicate = predicate;
    }

    public bool Test(A value) => predicate(value);

    public IContravariant<B> Contramap<B>(Func<B, A> f)
    {
        return new Predicate<B>(b => predicate(f(b)));
    }

    // Helper method for easier chaining
    public Predicate<B> ContramapTo<B>(Func<B, A> f)
    {
        return new Predicate<B>(b => predicate(f(b)));
    }
}

public class Person
{
    public int Age { get; set; }
    public string? Name { get; set; }
}


class Program
{
    static void Main()
    {
        // Base predicates (functions targeting bool)
        var isPositive = new Predicate<int>(x => x > 0);
        var isLongString = new Predicate<string>(s => s.Length > 5);

        // Contravariant mapping: int -> bool becomes string -> bool
        var stringIsPositive = isPositive.ContramapTo<string>(s => int.Parse(s));

        Console.WriteLine(isPositive.Test(5));            // True
        Console.WriteLine(isPositive.Test(-3));           // False
        Console.WriteLine(stringIsPositive.Test("10"));   // True
        Console.WriteLine(stringIsPositive.Test("-5"));   // False

        // Person example
        var isAdult = new Predicate<int>(age => age >= 18);
        var personIsAdult = isAdult.ContramapTo<Person>(p => p.Age);

        var alice = new Person { Age = 25, Name = "Alice" };
        var bob = new Person { Age = 16, Name = "Bob" };

        Console.WriteLine($"Alice is adult: {personIsAdult.Test(alice)}"); // True
        Console.WriteLine($"Bob is adult: {personIsAdult.Test(bob)}");     // False
   }
}

File: thinking-categorically/functor/contravariant-hom-functor/contravariant-hom-functor.ts
// Contravariant Hom functor: (- -> B) for fixed B
interface Contravariant<A> {
  contramap<B>(f: (b: B) => A): Contravariant<B>;
}

// Predicate contravariant functor: (A -> boolean)
class Predicate<A> implements Contravariant<A> {
  constructor(private predicate: (a: A) => boolean) {}

  test(value: A): boolean {
    return this.predicate(value);
  }

  contramap<B>(f: (b: B) => A): Predicate<B> {
    return new Predicate<B>((b: B) => this.predicate(f(b)));
  }
}

// Examples with target type boolean
const isPositive = new Predicate<number>(x => x > 0);
const isLongString = new Predicate<string>(s => s.length > 5);

// Contravariant mapping: transform number -> boolean into string -> boolean
// by first converting string to number
const stringToPositive = isPositive.contramap((s: string) => parseInt(s));

console.log(isPositive.test(5));           // true
console.log(isPositive.test(-3));          // false
console.log(stringToPositive.test("10"));  // true
console.log(stringToPositive.test("-5"));  // false

// Another example: Person -> boolean via Person -> number
interface Person { age: number; name: string; }
const isAdult = new Predicate<number>(age => age >= 18);
const personIsAdult = isAdult.contramap((p: Person) => p.age);

const alice: Person = { age: 25, name: "Alice" };
const bob: Person = { age: 16, name: "Bob" };

console.log(personIsAdult.test(alice)); // true
console.log(personIsAdult.test(bob));   // false
